src/commands/yubihsm/setup.rs

```diff
@@ -0,0 +1,334 @@
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+}
+
+impl Drop for RolePassword {
+    fn drop(&mut self) {
+        self.0.zeroize();
+    }
+}
+
+/// Derive a wrap key from the given BIP39 `Mnemonic`
+fn derive_wrap_key_from_mnemonic(mnemonic: &Mnemonic, key_id: object::Id) -> wrap::Key {
+    // Capabilities given to the initial wrap key:
+    // wrap/unwrap both data and objects
+    let wrap_key_capabilities = Capability::EXPORT_WRAPPED
+        | Capability::IMPORT_WRAPPED
+        | Capability::WRAP_DATA
+        | Capability::UNWRAP_DATA;
+
+    // Allow the initial wrap key to create new objects with any other
+    // capabilities.
+    //
+    // To prevent escalation of privilege, no roles except administrators
+    // will be given the capability to export other objects.
+    let wrap_key_delegated_capabilities = Capability::all();
+
+    // Make the wrap key accessible from all domains. This allows it to
+    // import and export objects from any domain.
+    let wrap_key_domains = Domain::all();
+
+    // Label to put on the wrap key
+    let wrap_key_label = create_object_label("primary");
+
+    // Includes the key ID in the derivation path, which allows us to derive
+    // other wrap keys from the same seed `Mnemonic` phrase in the event one
+    // has been compromised.
+    wrap::Key::from_bytes(
+        key_id,
+        &derive_secret_from_mnemonic(mnemonic, &[b"wrap", format!("{}", key_id).as_bytes()]),
+    )
+    .unwrap()
+    .label(wrap_key_label)
+    .capabilities(wrap_key_capabilities)
+    .delegated_capabilities(wrap_key_delegated_capabilities)
+    .domains(wrap_key_domains)
+}
+
+/// Derive secrets from the given BIP39 `Mnemonic` ala a BIP32 (hardened)
+/// derivation hierarchy.
+fn derive_secret_from_mnemonic(mnemonic: &Mnemonic, path: &[&[u8]]) -> [u8; KEY_SIZE] {
+    // Domain separate the toplevel of the derivation hierarchy ala BIP43
+    let mut seed_hmac = Hmac::<Sha512>::new_varkey(mnemonic.entropy()).unwrap();
+    seed_hmac.input(DERIVATION_VERSION);
+
+    let mut seed = [0u8; KEY_SIZE];
+    seed.copy_from_slice(&seed_hmac.result().code()[KEY_SIZE..]);
+
+    // Simplified BIP32-like hierarchical derivation
+    path.iter()
+        .enumerate()
+        .fold(seed, |mut parent_key, (i, elem)| {
+            let mut hmac = Hmac::<Sha512>::new_varkey(&parent_key).unwrap();
+            hmac.input(elem);
+
+            let hmac_result = hmac.result().code();
+            parent_key.zeroize();
+
+            let (secret_key, chain_code) = hmac_result.split_at(KEY_SIZE);
+            let mut child_key = [0u8; KEY_SIZE];
+
+            if i < path.len() - 1 {
+                // Use chain code for all but the last element
+                child_key.copy_from_slice(chain_code);
+            } else {
+                // Use secret key for the last element
+                child_key.copy_from_slice(secret_key);
+            }
+
+            child_key
+        })
```

Comment:
This looks very similar to bip32 derivation but much simpler. Should I spend some time to properly understand the subtle differences if any?

Reply:
The two main differences from BIP32-style hardened derivation at the moment are:

- No toplevel domain separation string (BIP32 uses `Bitcoin seed`). Instead it only includes a BIP43-style "purpose", which is set to `1`. Including this might be a good idea, as it would potentially allow people to cleanly reuse these words for other things if need be.
- Raw bytestring-based API instead of BIP32's child number system, which I consider to be incidental complexity arising from combining a symmetric key derivation scheme with a key blinding scheme.

Regarding the first issue, BIP32 curiously uses a fixed string as the HMAC *key*. This is fine owing to the unique properties of HMAC as a construction, but bad hygiene for PRFs in general (the key should be uniformly random, with the fixed strings used as PRF *inputs*). Minor nit in the construction of BIP32 in general.

Would there be merit to just using BIP32 outright? Maybe. Personally one of the things I like the most about the YubiHSMs is the ability to generate encrypted backups of keys generated within the HSMs, such that if a validator's "12 words" are ever compromised, they do not immediately give an attacker the ability to calculate a validator's private key. Instead, an attacker will need both that and an encrypted/wrapped backups of the private keys.

The other difficult thing about BIP32 is if we do plan to use it, I think it should probably be used correctly within a properly registered BIP43/BIP44 scheme (otherwise what's the point? and see also previous unpleasantness about registering numbers within derivation paths). That said, if people want to use BIP32 for something in the future, there's no reason they couldn't use it in parallel with the same mnemonic.

Personally I'm happy with something that looks enough like BIP32 to make it unconcerning from a security analysis POV, and simple enough it can be implemented in ~25 lines of code.